{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coverage tracks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Install cooltools env\n",
    "\n",
    "```\n",
    "mamba create -n pygenometracks -c bioconda -c conda-forge python=3.8\n",
    "mamba activate pygenometracks\n",
    "mamba install -c conda-forge -c bioconda pygenometracks=3.5\n",
    "mamba install -c anaconda ipykernel\n",
    "python -m ipykernel install --user --name=cooltools_env\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create tracks file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# where to save figures\n",
    "fig_dir=\"/home/anamaria/cluster/aelek/proj/scATAC_nvec_v2/Nematostella_scATAC/plots/Figures\"\n",
    "\n",
    "# annotation tracks\n",
    "gtf_fn=\"/home/anamaria/cluster/aelek/proj/scATAC_nvec_v2/genome/Nvec_vc1.1_long.annot.gtf\"\n",
    "bed_fn=\"/home/anamaria/cluster/aelek/proj/scATAC_nvec_v2/genome/Nvec_v4_merged_annotation_sort.bed\"\n",
    "rna_fwd=\"/home/anamaria/cluster/aelek/proj/RNAseq_nvec/Nvec_adult_polyp_fwd.bw\"\n",
    "rna_rev=\"/home/anamaria/cluster/aelek/proj/RNAseq_nvec/Nvec_adult_polyp_rev.bw\"\n",
    "rna_bws=$( echo $rna_fwd $rna_rev )\n",
    "chip_dir=\"/home/anamaria/cluster/aelek/proj/iChIP_nvec/signal\"\n",
    "chip_bws=$( ls ${chip_dir}/*bw )\n",
    "make_tracks_file --trackFiles ${gtf_fn} ${bed_fn} ${rna_bws} ${chip_bws} -o ${fig_dir}\"/fig1_tracks_annot.ini\"\n",
    "\n",
    "# broad cell types\n",
    "declare -A broad_cell_types\n",
    "broad_cell_types[cnidocyte]=\"#FF42FF\"\n",
    "broad_cell_types[ecto]=\"#51a0be\"\n",
    "broad_cell_types[EMS]=\"#bdf5bd\"\n",
    "broad_cell_types[gastro_circular_muscle]=\"#73b009\"\n",
    "broad_cell_types[gastro_parietal_muscle]=\"#8ceb10\"\n",
    "broad_cell_types[gastro]=\"#85c90e\"\n",
    "broad_cell_types[muscle]=\"#FFD700\"\n",
    "broad_cell_types[digestive_filaments]=\"#e33d3d\"\n",
    "broad_cell_types[precursors]=\"#BEBEBE\"\n",
    "broad_cell_types[epidermis]=\"#04ccd4\"\n",
    "broad_cell_types[neuronal]=\"#063cb9\"\n",
    "broad_cell_types[neuron_GATA_Islet]=\"#1175f0\"\n",
    "broad_cell_types[neuron_Pou4_FoxL2]=\"#101cde\"\n",
    "broad_cell_types[gland]=\"#ff6f08\"\n",
    "\n",
    "lvl=\"broad_cell_type\"\n",
    "bws_dir=\"/home/anamaria/cluster/aelek/proj/scATAC_nvec_v2/clustering/SEACells/Results/ArchRProj_Nvec_TSS4_frag200/ArchRProj/GroupBigWigs/\"${lvl}\n",
    "\n",
    "bws=\"\"\n",
    "col=\"\"\n",
    "cts=\"\"\n",
    "for i in ${!broad_cell_types[@]}\n",
    "do\n",
    "    ct=$i\n",
    "    cl=${broad_cell_types[$i]}\n",
    "    echo $ct $cl\n",
    "    bw=$( find \"${bws_dir}\" -name *${ct}-* )\n",
    "    bws=$( echo $bws $bw )\n",
    "    cts=$( echo $cts $ct )\n",
    "    col=$( echo $col $cl )\n",
    "done\n",
    "\n",
    "tracks=${fig_dir}\"/fig1_tracks_\"${lvl}\".ini\"\n",
    "\n",
    "make_tracks_file --trackFiles ${bws} -o ${tracks}\n",
    "\n",
    "# cell types\n",
    "declare -A cell_types\n",
    "cell_types[cnidocyte]=\"ff42ff\"\n",
    "cell_types[cnidocyte_precursors]=\"#f7abf7\"\n",
    "cell_types[ecto_pharynx]=\"#5bc0e8\"\n",
    "cell_types[ectoderm]=\"#51a0be\"\n",
    "cell_types[ecto_aboral]=\"#045170\"\n",
    "cell_types[EMS]=\"#bdf5bd\"\n",
    "cell_types[EMS_ecto_boundary]=\"#93dbce\"\n",
    "cell_types[gastro_circular_muscle_1]=\"#85c90e\"\n",
    "cell_types[gastro_circular_muscle_2]=\"#73b009\"\n",
    "cell_types[gastro_parietal_muscle]=\"#8ceb10\"\n",
    "cell_types[gastro_unknown_1]=\"#c1eb05\"\n",
    "cell_types[gastro_unknown_2]=\"#bde314\"\n",
    "cell_types[muscle_1]=\"#ffd700\"\n",
    "cell_types[muscle_2]=\"#f0e229\"\n",
    "cell_types[digestive_filaments_1]=\"#e33d3d\"\n",
    "cell_types[digestive_filaments_2]=\"#d10606\"\n",
    "cell_types[digestive_filaments_3]=\"#ad0303\"\n",
    "cell_types[epidermis_1]=\"#04ccd4\"\n",
    "cell_types[precursors_1]=\"#bebebe\"\n",
    "cell_types[precursors_2]=\"#8a8686\"\n",
    "cell_types[precursors_3]=\"#636363\"\n",
    "cell_types[NPC]=\"#808d91\"\n",
    "cell_types[neuron_GATA_Islet_1]=\"#0c82f7\"\n",
    "cell_types[neuron_GATA_Islet_2]=\"#1175f0\"\n",
    "cell_types[neuron_Pou4_FoxL2_1]=\"#101cde\"\n",
    "cell_types[neuron_Pou4_FoxL2_2]=\"#0b16bf\"\n",
    "cell_types[neuron_Pou4_FoxL2_3]=\"#2e39dd\"\n",
    "cell_types[neuronal]=\"#063cb9\"\n",
    "cell_types[gland]=\"#ff6f08\"\n",
    "\n",
    "lvl=\"cell_type\"\n",
    "bws_adult_dir=\"/home/anamaria/cluster/aelek/proj/scATAC_nvec_v2/clustering/SEACells/Results/ArchRProj_Nvec_TSS4_frag200/ArchRProj/GroupBigWigs/\"${lvl}\n",
    "bws_gastr_dir=\"/home/anamaria/cluster/aelek/proj/scATAC_nvec_gastrula/clustering/SEACells/Results/ArchRProj_Nvec_gastrula/ArchRProj/GroupBigWigs/\"${lvl}\n",
    "\n",
    "bws=\"\"\n",
    "col=\"\"\n",
    "cts=\"\"\n",
    "for i in ${!cell_types[@]}\n",
    "do\n",
    "    ct=$i\n",
    "    cl=${cell_types[$i]}\n",
    "    echo $ct $cl\n",
    "    bw=$( find \"${bws_adult_dir}\" -name *${ct}-* )\n",
    "    bws=$( echo $bws $bw )\n",
    "    cts=$( echo $cts $ct )\n",
    "    col=$( echo $col $cl )\n",
    "done\n",
    "make_tracks_file --trackFiles ${bws} -o ${fig_dir}\"/fig1_tracks_\"${lvl}\"_adult.ini\"\n",
    "\n",
    "bws=\"\"\n",
    "col=\"\"\n",
    "cts=\"\"\n",
    "for i in ${!cell_types[@]}\n",
    "do\n",
    "    ct=$i\n",
    "    cl=${cell_types[$i]}\n",
    "    echo $ct $cl\n",
    "    bw=$( find \"${bws_gastr_dir}\" -name *${ct}-* )\n",
    "    bws=$( echo $bws $bw )\n",
    "    cts=$( echo $cts $ct )\n",
    "    col=$( echo $col $cl )\n",
    "done\n",
    "make_tracks_file --trackFiles ${bws} -o ${fig_dir}\"/fig1_tracks_\"${lvl}\"_gastrula.ini\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the files above, I will manually assemble the final ini file for the figure: `fig1_tracks_ini` and then save plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "declare -A marker_regions\n",
    "marker_regions[\"GATA\"]=\"NC_064043.1:5890694-5915400\"\n",
    "marker_regions[\"Elav\"]=\"NC_064034.1:3163000-3178000\"\n",
    "marker_regions[\"Pou4\"]=\"NC_064043.1:10388438-10402026\" # bad H3K4me3\n",
    "marker_regions[\"FoxA\"]=\"NC_064035.1:7886656-7907856\"\n",
    "marker_regions[\"Fgfa1\"]=\"NC_064043.1:7805055-7811856\" # bad H3K4me3\n",
    "marker_regions[\"GABRA\"]=\"NC_064035.1:11632899-11660053\" # bad H3K4me3\n",
    "marker_regions[\"NVHD033-ANTP\"]=\"NC_064041.1:14326600-14335300\" # could-be-worse H3K4me3\n",
    "marker_regions[\"HMGbox_SOX1_2_3_15\"]=\"NC_064042.1:10340550-10350000\" \n",
    "\n",
    "declare -A marker_genes\n",
    "marker_genes[\"GATA\"]=\"Nvec_vc1.1_XM_032367288.2\"\n",
    "marker_genes[\"Elav\"]=\"Nvec_vc1.1_XM_048724864.1\"\n",
    "marker_genes[\"Pou4\"]=\"Nvec_vc1.1_XM_032363992.2\"\n",
    "marker_genes[\"FoxA\"]=\"Nvec_vc1.1_XM_001634505.3\"\n",
    "marker_genes[\"Fgfa1\"]=\"Nvec_vc1.1_XM_032367336.2\"\n",
    "marker_genes[\"GABRA\"]=\"Nvec_vc1.1_XM_048723418.1\"\n",
    "marker_genes[\"NVHD033-ANTP\"]=\"Nvec_vc1.1_XM_032379688.2\"\n",
    "marker_genes[\"HMGbox_SOX1_2_3_15\"]=\"Nvec_vc1.1_XM_001632947.3\"\n",
    "\n",
    "# scATAC with selected bulk tracks\n",
    "tracks=${fig_dir}\"/fig1_tracks_ini\"\n",
    "for i in ${!marker_genes[@]}\n",
    "do\n",
    "    gene=$i\n",
    "    cord=${marker_regions[$gene]}\n",
    "    plot=${fig_dir}/pyGenomeTracks_${gene}.svg\n",
    "    pyGenomeTracks --tracks ${tracks} --region ${cord} --title ${gene} -o ${plot}\n",
    "done\n",
    "\n",
    "# scRNA\n",
    "tracks=${fig_dir}\"/tracks_scRNA.ini\"\n",
    "for i in ${!marker_genes[@]}\n",
    "do\n",
    "    gene=$i\n",
    "    cord=${marker_regions[$gene]}\n",
    "    plot=${fig_dir}/pyGenomeTracks_${gene}_scRNA.svg\n",
    "    pyGenomeTracks --tracks ${tracks} --region ${cord} --title ${gene} -o ${plot}\n",
    "done\n",
    "\n",
    "# loop over two arrays and make plots\n",
    "tracks=${fig_dir}\"/fig1_tracks.ini\"\n",
    "for i in ${!marker_genes[@]}\n",
    "do\n",
    "    gene=$i\n",
    "    cord=${marker_regions[$gene]}\n",
    "    echo\n",
    "    echo '# # # # # # # # # '$gene : $cord'# # # # # # # # # '\n",
    "    echo\n",
    "    plot=${fig_dir}/pyGenomeTracks_${gene}.svg\n",
    "    pyGenomeTracks --tracks ${tracks} --region ${cord} --title ${gene} -o ${plot}\n",
    "done\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "2.7.18"
  },
  "vscode": {
   "interpreter": {
    "hash": "e7370f93d1d0cde622a1f8e1c04877d8463912d04d973331ad4851f04de6915a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
